source/particle/property/viscoplastic_strain_invariants.cc

```diff
@@ -117,45 +120,11 @@ namespace aspect
             material_inputs.composition[0][i] = solution[this->introspection().component_indices.compositional_fields[i]];
           }
 
-        // Find out plastic yielding by calling function in material model.
-        const MaterialModel::ViscoPlastic<dim> &viscoplastic
-          = Plugins::get_plugin_as_type<const MaterialModel::ViscoPlastic<dim>>(this->get_material_model());
-
-        const bool plastic_yielding = viscoplastic.is_yielding(material_inputs);
-
-
-        /* Next take the integrated strain invariant from the prior time step. When
-         * there are two fields (plastic and viscous), this assumes plastic
-         * strain is always in data position one and viscous strain in position two.
-         * In this case old_strain will first be given the plastic strain, and then,
-         * if there is no plastic yielding it will update to the viscous strain instead.
-         */
-        auto &data = particle->get_properties();
-        double old_strain = data[data_position];
-        if (n_components == 2 && plastic_yielding == false)
-          old_strain = data[data_position+(n_components-1)];
-
-
-        // Calculate strain rate second invariant
-        const double edot_ii = std::sqrt(std::fabs(second_invariant(deviator(material_inputs.strain_rate[0]))));
+        // Evaluate directly in the viscoplastic material model and modify the reaction outputs
+        this->get_material_model().evaluate (material_inputs,material_outputs);
 
-        // New strain is the old strain plus dt*edot_ii
-        const double new_strain = old_strain + dt*edot_ii;
-
-
-        /* Once we know whether the particle underwent plastic, viscous, or
-         * total strain assign the new strain to the correct data position.
-         * NOTE: This assumes that total strain cannot be used in combination with the
-         * other fields. If this changes in the future, this will need to be updated.
-         * */
-        if (this->introspection().compositional_name_exists("plastic_strain") && plastic_yielding == true)
-          data[data_position] = new_strain;
-
-        if (this->introspection().compositional_name_exists("viscous_strain") && plastic_yielding == false)
-          data[data_position+(n_components-1)] = new_strain;
-
-        if (this->introspection().compositional_name_exists("total_strain"))
-          data[data_position] = new_strain;
+        for (unsigned int i = 0; i < SymmetricTensor<2,dim>::n_independent_components ; ++i)
+          particle->get_properties()[data_position + i] += material_outputs.reaction_terms[0][i];
```

Comment:
I dont understand this part. This particle property has between 0 and 3 components (take a look at lines 58 and 65), but not `SymmetricTensor<2,dim>::n_independent_components` (=3 for dim==2 and =6 for dim==3). You probably want to assign reaction terms of certain indices (the indices of the strain components) to the particle property components `0` to `n_components-1`. To find out the indices of the reaction terms, you can use the same approach as in lines 82-89. 

After that works please check our existing test cases and make sure the results with particles are similar to the results with compositional fields (they will likely not be completely identical, but should be close enough).

Reply:
Whoops, yes it should be n_components-1 to match, I had copied that line over from elastic stress and forgot to change it.

Reply:
@erinheilman - to follow-up on the comment from  @gassmoeller, you can get the compositional index for a certain strain field using the field names. 

For example, to get the index for plastic strain you could use:
`const int plastic_strain_index = this->introspection().compositional_index_for_name("plastic_strain")`

This could then be used as follows (I think):
```
if (this->introspection().compositional_name_exists("plastic_strain")
    particle->get_properties()[data_position] += material_outputs.reaction_terms[0][plastic_strain_index];
```

The logic in the old version of this file (lines 154-158) would be needed for the viscous and total strain.

Reply:
So essentially do this,
        const int plastic_strain_index = this->introspection().compositional_index_for_name("plastic_strain") 
        const int viscous_strain_index = this->introspection().compositional_index_for_name("viscous_strain")
        const int total_strain_index = this->introspection().compositional_index_for_name("total_strain")
        
        if (this->introspection().compositional_name_exists("plastic_strain")
          particle->get_properties()[data_position] += material_outputs.reaction_terms[0][plastic_strain_index];
        
        if (this->introspection().compositional_name_exists("viscous_strain")
          particle->get_properties()[data_position] += material_outputs.reaction_terms[0][viscous_strain_index];

        if (this->introspection().compositional_name_exists("total_strain")
          particle->get_properties()[data_position] += material_outputs.reaction_terms[0][total_strain_index]; 
instead of the for loop to update the reaction terms?

Reply:
@erinheilman - In case you did not see it, @gassmoeller provided the right structure in his last review.